src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java

```diff
@@ -0,0 +1,336 @@
+            boolean moreReplicatesNeeded = replicatesService.moreReplicatesNeeded(chainTaskId,
+                    task.getNumWorkersNeeded(), task.getMaxExecutionTime());
+
+            if (blockNumberAvailable && !hasWorkerAlreadyParticipated && moreReplicatesNeeded) {
+                replicatesService.addNewReplicate(chainTaskId, walletAddress);
+                workerService.addChainTaskIdToWorker(chainTaskId, walletAddress);
+                return replicatesService.getReplicate(chainTaskId, walletAddress);
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    public List<InterruptedReplicateModel> getInterruptedReplicates(long blockNumber, String walletAddress) {
+
+        List<String> chainTaskIdList = workerService.getChainTaskIds(walletAddress);
+        List<Task> tasksWithWorkerParticipation = taskService.getTasksByChainTaskIds(chainTaskIdList);
+        List<InterruptedReplicateModel> interruptedReplicates = new ArrayList<>();
+
+        for (Task task : tasksWithWorkerParticipation) {
+
+            Optional<Replicate> oReplicate = replicatesService.getReplicate(
+                    task.getChainTaskId(), walletAddress);
+            if (!oReplicate.isPresent()) continue;
+
+            Replicate replicate = oReplicate.get();
+
+            boolean isRecoverable = replicate.isRecoverable();
+            if (!isRecoverable) continue;
+
+            RecoveryAction recoveryAction = getAppropriateRecoveryAction(task, replicate, blockNumber);
+            if (recoveryAction == null) continue;
+
+            // generate contribution authorization
+            Optional<ContributionAuthorization> authorization = signatureService.createAuthorization(
+                    walletAddress, task.getChainTaskId(), TeeUtils.isTrustedExecutionTag(task.getTag()));
+
+            InterruptedReplicateModel interruptedReplicate = InterruptedReplicateModel.builder()
+                    .contributionAuthorization(authorization.get())
+                    .recoveryAction(recoveryAction)
+                    .build();
+
+            // change replicate status
+            replicatesService.updateReplicateStatus(task.getChainTaskId(), walletAddress,
+                    ReplicateStatus.RECOVERING, ReplicateStatusModifier.POOL_MANAGER);
+
+            interruptedReplicates.add(interruptedReplicate);
+        }
+
+        return interruptedReplicates;
+    }
+
+    public RecoveryAction getAppropriateRecoveryAction(Task task, Replicate replicate, long blockNumber) {
+        ChainReceipt chainReceipt = new ChainReceipt(blockNumber, "");
+
+        if (task.inContributionPhase()) {
+            return recoverReplicateInContributionPhase(task, replicate, chainReceipt);
+        }
+
+        if (task.getCurrentStatus().equals(TaskStatus.CONTRIBUTION_TIMEOUT)) {
+            return RecoveryAction.ABORT_CONTRIBUTION_TIMEOUT;
+        }
+
+        if (task.getCurrentStatus().equals(TaskStatus.CONSENSUS_REACHED)) {
```

Comment:
Those two ifs can be combined into one.

Reply:
I thought it would be more readable, but it's okey let's change it